src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5074,7 +5074,26 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                 SelectWithUnderlyingError(tree, err)
                 return tree
               case SilentResultValue((qual, pre)) =>
-                (stabilize(qual, pre, mode, pt), None)
+                val tree = stabilize(qual, pre, mode, pt)
+                if (tree.symbol.isStable && treeInfo.isExprSafeToInline(qual)) {
+                  tree
+                } else {
+                  def deconstResult(tp: Type): Type =
+                    tp match {
+                    case tp @ NullaryMethodType(res) =>
+                      val res1 = res.deconst
+                      if (res eq res1) tp else NullaryMethodType(res1)
+                    case tp @ MethodType(params, res) =>
+                      val res1 = deconstResult(res)
+                      if (res eq res1) tp else MethodType(params, res1)
+                    case tp @ PolyType(params, res) =>
+                      val res1 = deconstResult(res)
+                      if (res eq res1) tp else PolyType(params, res1)
+                    case _ => tp.deconst
+                  }
+                  tree.modifyType(deconstResult)
+                }
+                (tree, None)
```

Comment:
I've also been [looking](https://github.com/scala/scala/compare/2.13.x...retronym:topic/constant-folder-unary?expand=1) at this line of code in the context of figuring out why `final val x = ~0` isn't constant folding as expected. I think there is a missing call to `constfold` here. It seems like `stabilize`, the new code here, and `constfold` are all testing similar properties `e.g. sym.isStable`. I think there is opportunity to consolidate some of this, but I'm still trying to get a decent mental map of this terrain, so I don't have a concrete suggestion yet.